kong/plugins/aws-lambda/handler.lua

```diff
@@ -25,32 +24,77 @@ function AWSLambdaHandler:new()
   AWSLambdaHandler.super.new(self, "aws-lambda")
 end
 
-local function retrieve_parameters()
-  ngx_req_read_body()
+local function retrieve_parameters(content_type, conf)
   local body_parameters, err
-  local content_type = ngx_req_get_headers()[CONTENT_TYPE]
-  if content_type and string_find(content_type:lower(), "multipart/form-data", nil, true) then
+  if content_type and string_find(content_type, "multipart/form-data", nil, true) then
     body_parameters = Multipart(ngx_req_get_body_data(), content_type):get_all()
-  elseif content_type and string_find(content_type:lower(), "application/json", nil, true) then
+  elseif content_type and string_find(content_type, "application/json", nil, true)
+         and not conf.request_body then
     body_parameters, err = cjson.decode(ngx_req_get_body_data())
     if err then
       body_parameters = {}
     end
-  else
+  elseif content_type and string_find(content_type, "application/x-www-form-urlencoded", nil, true) then
     body_parameters = public_utils.get_post_args()
+  else
+    body_parameters = {}
   end
 
   return utils.table_merge(ngx_req_get_uri_args(), body_parameters)
 end
 
+local function retrieve_body(content_type)
+  local err, body
+  local raw_body = ngx_req_get_body_data()
+  if raw_body == nil then return nil end
+
+  if content_type and string_find(content_type, "application/json", nil, true) then
+    body, err = cjson.decode(raw_body)
+    if err then
+      return nil
+    end
+  elseif content_type and ( string_find(content_type, "application/xml", nil, true)
+    or  string_find(content_type, "application/soap+xml", nil, true)
+    or  string_find(content_type, "text/", nil, true) )
+  then
+    body = raw_body
+  else
+    body = ngx_encode_base64(raw_body)
+  end
+  return body
+end
+
 function AWSLambdaHandler:access(conf)
   AWSLambdaHandler.super.access(self)
 
-  local bodyJson = cjson.encode(retrieve_parameters())
+  local content_type = ngx.var.http_content_type
+  if content_type then
+     content_type = content_type:lower()
+  end
+
+  ngx_req_read_body()
 
-  local host = string.format("lambda.%s.amazonaws.com", conf.aws_region)
-  local path = string.format("/2015-03-31/functions/%s/invocations",
+  local body_parameters = retrieve_parameters(content_type, conf)
```

Comment:
Also, I'm not sure I feel comfortable with mixing the new values from this PR with the request body, at least at the same root level. For semantics but also due to possible conflicts and overriding a previous body parameter.

I feel like the proper behavior would be:

Instead:
```lua
local upstream_body = {
  request_body_args    = retrieve_parameters(), -- replaced once #2503 gets merged
  request_body_raw     = retrieve_body(content_type),
  request_http_headers = ngx_req_get_headers(),
  request_http_method  = var.request_method,
  request_uri          = var.request_uri,
}
```

Sadly, this would be a breaking change for this plugin, but I do not see a possibility of this being merged as-is, for the reasons described above...